### PR TITLE
Rewrite logging to structured log format

### PR DESCRIPTION
**Pull request by bergzand**
This commit rewrites the logging in the application to structured log
format. A pretty console printing and direct systemd-journal are available.
The logging can be switched via the MURDOCK_LOG_OUTPUT environment
variable. It defaults to `console` for development. Setting it to
`journal` sets it to output to the systemd journal.

### CODE REVIEW

**Comment by bergzand** [comment]
Also needs #35

**Comment by bergzand** [comment]
Rebased!

**Comment by bergzand** [comment]
This PR includes #36 and should be rebased later when that PR is merged

**Comment by aabadie** [comment]
Please rebase (#36 is merged)

File: murdock/main.py [hunk 1]
```diff
@@ -79,11 +80,11 @@
     StaticFiles(directory=GLOBAL_CONFIG.work_dir, html=True, check_dir=False),
     name="results",
 )
+murdock.instrumentator.instrument(app).expose(app)
```
Comment (aabadie) [inline murdock/main.py:83]: Why is this moved outside of the startup event function ? Is it because of the way the logger is setup ?

Comment (bergzand) [reply murdock/main.py:83]: I triggered an issue with starlette that it doesn't allow for adding middleware anymore after starting the app.

Comment (bergzand) [reply murdock/main.py:83]: Not sure btw what triggered this. Locally I have starlette-0.24.0

**Comment by aabadie** [comment]
Hmm, this needs a rebase, now that #32 is in.

**Comment by bergzand** [comment]
> Hmm, this needs a rebase, now that #32 is in.

And another rebase

File: murdock/log.py [hunk 2]
```diff
@@ -1,20 +1,73 @@
...
+    },
+    "handlers": {
+        "console": {
+            "formatter": "rich",
+            "class": "logging.StreamHandler",
+            "stream": "ext://sys.stderr",
+        },
+        "journal": {
+            "formatter": "logfmt",
+            "()": journal.JournaldLogHandler,
```
[the remark this answers is not in the document]
Comment (bergzand) [reply murdock/log.py:58]: Good catch, do you have a suggestion for a clean fix here?

[the remark this answers is not in the document]
Comment (bergzand) [reply murdock/log.py]: Added a fixup with this. Tested and works here with and without `cysystemd` installed. It does throw a rather blunt exception if you try to log to an unavailable log handler :)

File: murdock/job.py [hunk 3]
```diff
@@ -235,8 +242,12 @@ async def extend_job_output(self, line):
                 json.dumps({"cmd": "output", "uid": self.uid, "line": line})
             )
 
+    def logger(self):
```
Comment (aabadie) [inline murdock/job.py:245]: If this really used ? Could be decorated with `@property` I think. Then all `self._logger` could be replaced with `self.logger`.

File: murdock/main.py (hunk 1 above)
Comment (aabadie) [reply murdock/main.py:83]: Ok, that's not really an issue anyway, let's keep it like this.

**Comment by aabadie** [comment]
Hmm, the CI is complaining because of mypy. I don't know what the problem is. Otherwise the changes looks ok in general. Once mypy is fixed that can be merged I think.

**Comment by aabadie** [comment]
And please squash!

File: murdock/job.py (hunk 3 above)
Comment (bergzand) [reply murdock/job.py:245]: There are a few cases in `murdock.py` where it is logging information about a job, so there it is easier to construct a new [bound logger](https://www.structlog.org/en/stable/bound-loggers.html) based on the job logger to include the job's context in the log line.

I agree that this could be cleaner, let me see what I can come up with.

Comment (bergzand) [reply murdock/job.py]: I rewrote this a bit, removing this function and adding a `logging_context` property here to expose the logging context that can be directly used with `LOGGER.bind` calls outside the `job` class

File: murdock/job.py [hunk 4]
```diff
@@ -61,24 +61,34 @@ def __init__(
         self.output_url = os.path.join(
             GLOBAL_CONFIG.base_url, "results", self.uid, "output"
         )
+        self._logger_context = {
+            "job": str(self.uuid),
+            "job_description": self,
```
Comment (aabadie) [inline murdock/job.py:66]: I would find this nicer to use `str(self)`, this way it's explicit that the `__repr__` is used (IIUC).

File: murdock/job.py [hunk 5]
```diff
@@ -235,8 +245,17 @@ async def extend_job_output(self, line):
                 json.dumps({"cmd": "output", "uid": self.uid, "line": line})
             )
 
+    @property
+    def logging_context(self):
+        # Copy over the dict
+        return dict(self._logger_context)
+
+    def clone_logger(self, **context):
```
Comment (aabadie) [inline murdock/job.py:253]: I can't find where this is used (via the github UI)

File: murdock/job.py (hunk 4 above)
Comment (bergzand) [reply murdock/job.py]: You're right, added.

**Review by aabadie — APPROVED** [review]
All good!
